### PR TITLE
Take binary version in Downgrade/Upgrade Test

### PR DESCRIPTION
**Pull request by tjungblu**
This fix avoids the assumption of knowing the current version of the binary. We can query the binary with the version flag to get the actual version of the given binary we upgrade and downgrade to. The respectively reported versions should match what is returned by the version endpoint.

Signed-off-by: Thomas Jungblut <tjungblu@redhat.com>

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: tests/e2e/cluster_downgrade_test.go [hunk 1]
```diff
@@ -235,3 +242,22 @@ func getMemberVersionByCurl(cfg *e2e.EtcdProcessClusterConfig, member e2e.EtcdPr
...
+	if err != nil {
+		return nil, fmt.Errorf("could not find binary version from %s, err: %w", binaryPath, err)
+	}
+
+	for _, line := range lines {
+		if strings.HasPrefix(line, "etcd Version:") {
+			versionString := strings.TrimSpace(strings.SplitAfter(line, ":")[1])
+			if strings.HasSuffix(versionString, "-alpha.0") {
+				versionString = strings.Replace(versionString, "-alpha.0", "", 1)
+			}
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:257]:
```suggestion
```

File: tests/e2e/cluster_downgrade_test.go [hunk 2]
```diff
@@ -47,13 +48,19 @@ func testDowngradeUpgrade(t *testing.T, clusterSize int) {
...
 
-	currentVersion := semver.New(version.Version)
-	lastVersion := semver.Version{Major: currentVersion.Major, Minor: currentVersion.Minor - 1}
+	currentVersion, err := getVersionFromBinary(currentEtcdBinary)
+	require.NoError(t, err)
+	lastVersion, err := getVersionFromBinary(lastReleaseBinary)
+	require.NoError(t, err)
 
-	currentVersion.PreRelease = ""
+	require.Equalf(t, lastVersion.Minor, currentVersion.Minor-1, "unexpected minor version difference")
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:56]: Nice!

File: tests/e2e/cluster_downgrade_test.go [hunk 3]
```diff
@@ -47,13 +48,19 @@ func testDowngradeUpgrade(t *testing.T, clusterSize int) {
...
 	currentVersionStr := currentVersion.String()
 	lastVersionStr := lastVersion.String()
 
+	lastVersionZero := semver.New(lastVersionStr)
+	lastVersionZero.Patch = 0
+	lastVersionZeroStr := lastVersionZero.String()
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:62]:
```suggestion
	lastVersionZero := version.Cluster(lastVersionStr)
```

File: tests/e2e/cluster_downgrade_test.go [hunk 4]
```diff
@@ -73,9 +80,9 @@ func testDowngradeUpgrade(t *testing.T, clusterSize int) {
 	t.Log("Downgrade enabled, validating if cluster is ready for downgrade")
 	for i := 0; i < len(epc.Procs); i++ {
 		validateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{
-			Cluster: lastVersionStr,
+			Cluster: lastVersionZeroStr,
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:83]: Maybe we could rename it to `lastClusterVersion`? 
We could have just a separate variable for `lastStorageVersion`.

File: tests/e2e/cluster_downgrade_test.go (hunk 1 above)
Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:257]: true, I try to remember why I added this last week. I thought this specific version wouldn't parse correctly. I'll try this pre-release trick again tomorrow.

Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:257]: works, using that instead.

File: tests/e2e/cluster_downgrade_test.go (hunk 3 above)
Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:62]: that doesn't work quite as well:
> {"error": "expect etcdcluster version 3.5, but got 3.5.0"}

The cluster version splits the patch release off

File: tests/e2e/cluster_downgrade_test.go (hunk 4 above)
Comment (tjungblu) [reply tests/e2e/cluster_downgrade_test.go:83]: renamed

**Comment by tjungblu** [comment]
very cool, let me know if I should adjust the GH action workflow as well.

File: tests/e2e/cluster_downgrade_test.go (hunk 3 above)
Comment (serathius) [reply tests/e2e/cluster_downgrade_test.go]: Ok, you might need to go through `semver.New().String()`, but let's use `version.Cluster` to get cluster version.

File: tests/e2e/cluster_downgrade_test.go [hunk 5]
```diff
@@ -47,13 +48,22 @@ func testDowngradeUpgrade(t *testing.T, clusterSize int) {
...
 	currentVersion.PreRelease = ""
+
+	lastVersion, err := getVersionFromBinary(lastReleaseBinary)
+	require.NoError(t, err)
+
+	require.Equalf(t, lastVersion.Minor, currentVersion.Minor-1, "unexpected minor version difference")
 	currentVersionStr := currentVersion.String()
 	lastVersionStr := lastVersion.String()
 
+	lastClusterVersion := semver.New(lastVersionStr)
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:63]:
```suggestion
	lastClusterVersion := semver.New(version.Cluster(lastVersionStr))
```

File: tests/e2e/cluster_downgrade_test.go [hunk 6]
```diff
@@ -47,13 +48,22 @@ func testDowngradeUpgrade(t *testing.T, clusterSize int) {
...
+	lastClusterVersion.Patch = 0
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:64]:
```suggestion
	lastClusterVersion := semver.New(version.Cluister(lastVersionStr))
```

File: tests/e2e/cluster_downgrade_test.go [hunk 7]
```diff
@@ -47,13 +48,22 @@ func testDowngradeUpgrade(t *testing.T, clusterSize int) {
...
+	lastClusterVersionStr := lastClusterVersion.String()
```
Comment (serathius) [inline tests/e2e/cluster_downgrade_test.go:65]:
```suggestion
	lastClusterVersionStr := semver.New(version.Cluster(lastVersionStr)).String()
```
Will this work?

**Review by serathius — APPROVED** [review]

File: tests/e2e/cluster_downgrade_test.go (hunk 7 above)
Comment (serathius) [reply tests/e2e/cluster_downgrade_test.go:65]: It's more of a nit, so let's we can skip it in this PR.

**Comment by serathius** [comment]
> very cool, let me know if I should adjust the GH action workflow as well.

Feel free to bump the old version used in upgrade/downgrade test.

**Pull request merged**